src/actions/collapseContext.ts

```diff
@@ -76,17 +83,52 @@ const collapseContext = (state: State, { at }: Options) => {
       newValue: createId(), // unique value
       path: simplePath,
     }),
-    // Sort parent context if sort preference exists
+    // Sort parent context if sort preference collapses
     sortId ? sort(parentId) : null,
+
     // outdent each child
-    ...children.map(
-      (child, i) => (state: State) =>
-        moveThought(state, {
+    ...children.map(child => (state: State) => {
+      // If we're inserting into a sorted context, short-circuit and use the sorted rank
+      if (isSortedContext)
+        return moveThought(state, {
           oldPath: appendToPath(simplePath, child.id),
           newPath: appendToPath(parentOf(simplePath), child.id),
-          newRank: sortId ? getSortedRank(state, parentId, child.value) : rankStart + rankIncrement * i,
-        }),
-    ),
+          newRank: getSortedRank(state, parentId, child.value),
+        })
+
+      const isMetaAttribute = isAttribute(child.value)
+      const firstNonMetaChild = findAnyChild(state, parentId, thought => !isAttribute(thought.value))
+      const insertMetaBeforePath = firstNonMetaChild
+        ? appendToPath(parentOf(simplePath), firstNonMetaChild.id)
+        : simplePath
```

Comment:
It's not necessary to calculate `firstNonMetaChild` or `insertMetaBeforePath` when `isMetaAttribute` is false.